library/kotlin/src/io/envoyproxy/envoymobile/Result.kt

```diff
@@ -0,0 +1,105 @@
+package io.envoyproxy.envoymobile
+
+/**
+ * The result of a request
+ */
+sealed class Result {
+
+  /**
+   * This method transforms a given response or failure to a specified type which
+   * the user can define. Usages of this can be:
+   *
+   * val result: Result = ...
+   *
+   * // To throw an exception when a failure occurs
+   * result.fold({ response -> response.body }, { failure -> throw Exception(failure.cause) })
+   *
+   * // To return a default value of 0 when a failure occurs
+   * result.fold({ response -> 1 }, { failure -> 0 })
```

Comment:
Sorry, this is my confused eyes as a kotlin ignorant, but how does the user interact with the result. Do they write complex folds where the logic is after `->` for response, and failure? And they have access to the Response and the NetworkError.

Reply:
I have an example usage of this up in the description but you do have the right idea